README.md

![](https://cloud.githubusercontent.com/assets/2152766/16178722/30c3d28c-3648-11e6-8361-84f04113af4b.gif)

[![Build Status](https://travis-ci.org/pyblish/pyblish-qml.svg?branch=master)](https://travis-ci.org/pyblish/pyblish-qml)

### Pyblish QML

The Pyblish QML project provides a graphical frontend to [Pyblish](http://pyblish.com).

<br>
<br>
<br>

### Requirements

The Pyblish QML **client** depends on [pyblish-base](https://github.com/pyblish/pyblish-base) and Python 2 or 3, whereas the **server** requires PyQt5.

- **Any platform** 
 
 PyQt5 is available on PyPI, but only for Python 3.5. For any other distribution, see below.

 ```bash
$ pip install PyQt5
```

- **Windows**

 On Windows, you may download and install the binary distribution from the [Riverbank Software website](https://www.riverbankcomputing.com/software/pyqt/download5).

- **Debian**

 Like most Linux distribution, Debian will have PyQt5 available via its native package manager.

 ```bash
$ apt-get install python3-pyqt5 python3-pyqt5.qtquick
```

 See `yum` for Red Hat based distributions.

- **MacOS**

 Finally, brew provides a pre-compiled distribution for Python 3 as well.

 ```bash
$ brew install pyqt5
```

<br>
<br>
<br>

### Installation

```bash
$ pip install pyblish-qml
```

Test out the installation with..

```bash
$ python -m pyblish_qml --demo
```

<br>
<br>
<br>

### Usage

Run Pyblish QML from any terminal, then show it from your favourite digital content creation software.

**Server**

```bash
# From any terminal
$ python -m pyblish_qml
```

**Client**

```python
# From your favourite DCC
import pyblish_qml
pyblish_qml.show()
```

<br>
<br>
<br>

### Under the Hood

Pyblish QML runs as an independent process on your computer, and communicates with your host via interprocess communication through remote-procedure calls (RPC).

It uses the standard [`xmlrpc`](https://docs.python.org/2/library/xmlrpclib.html) library and default to responding to calls via port number `9090`.

```python
>>> from xmlrpclib import ServerProxy
>>> proxy = ServerProxy("http://127.0.0.1:9090", allow_none=True)
>>> proxy.ping()
{'message': 'Hello, whomever you are'}
```

When you show Pyblish QML from within a host, you are effectively making an IPC connection and calling `show` with parameters relative the currently running process.

```python
>>> proxy.show(9001, {})
# GUI appears
```

The `9001` refers to the port number through which Pyblish QML may reach whomever is asking it to show. When calling `pyblish_qml.show()`, a listener is automatically started for you to receive calls from Pyblish QML; unless one is already running.

```python
def show():
    if not listener_started:
        start_listener()

    proxy.show()
```

You can manually start such a listener by calling..

```python
import pyblish_qml
pyblish_qml.install()
```

This needs only happen once. To shutdown the listener, you may call..

```python
pyblish_qml.uninstall()
```

<br>
<br>
<br>

### Differences to [Pyblish Lite](https://github.com/mottosso/pyblish-lite)

Pyblish QML fills the same gap as Pyblish Lite, with a few notable differences.

**Pros**

- Asynchronous operation - use the GUI during intense processing
- Faster startup time - it's running before you are
- Smoother visuals - animations galore
- Inspect individual items - tens of instances, hundreds of plug-ins? no problem
- Filter terminal via keyword search - thousands of log entries? no problem

**Cons**

- Requires PyQt5 (and either Python 2 or 3)
- Supports only one publish at a time, for one logged on user at a time ([#199](https://github.com/pyblish/pyblish-qml/issues/199))

Development wise, Pyblish QML is written in.. you guessed it, QML. Whereas Pyblish Lite is written using classig widgets. QML is a new graphical user interface language for OpenGL developed by the same group, Qt.

<br>
<br>
<br>

### Testing

Tests are automatically run at each commit to GitHub via Travis-CI. You can run these tests locally via Docker too.

```bash
$ git clone https://github.com/pyblish/pyblish-qml.git
$ cd pyblish-qml
$ docker build -t pyblish/pyblish-qml .
$ docker run --rm -v $(pwd):/pyblish-qml pyblish/pyblish-qml
# Doctest: pyblish_qml.models.Item ... ok
# Doctest: pyblish_qml.util.ItemList ... ok
# Reset works ... ok
# Publishing works ... ok
# ...
# util.chain works with lambdas ... ok
# 
# ----------------------------------------------------------------------
# Ran 20 tests in 1.430s
# 
# OK
```


CHANGES
Changelog
=========

Version 0.7.0
-------------

- Python 3 support (#191)
- Consolidation of Pyblish RPC (#202)
- Consolidation of Pyblish Integration (#203)

Version 0.6.0
-------------

- Feature: Action icons (#166)
- Feature: Added `ping()` to Proxy (#182)
- Enhancement: More robust handling of items via id's (#181)
- Enhancement: More debug info in terminal (#171)
- Enhancement: Syntax updated to Pyblish 1.3
- Bugfix: Toggle instance doesn't update compatibility view correctly (#173)
- Bugfix: Actions Availability (#172)
- 

Version 0.5.0
-------------

- Feature: Added support for Collaborative Collection
- Feature: `publish` data on Instance set when toggling in GUI (#125)
- Enhancement: Emitting "published" and "validated" signals (#161)
- Enhancement: Improved emitting of signals (#125)
- Enhancement: refresh() is confusing (#167)
- Enhancement: Python 3 compatibility (#159)
- Bugfix: Context doesn't refresh between plug-ins (#157)
- Bugfix: Inactive InstancePlugin set as CollectorOrder still runs (#169)
- Bugfix: Plugins display when no compatible families are present (#154)

Version 0.4.0
-------------

- Feature: Callbacks (#148)
- Enhancement: Collectors visible during reset (#133)
- Enhancement: Animated feedback on currently running items
- Enhancement: More visible indicator of which is currently running
- Optimisation: Start-up performance; the Window pops up faster thanks for not having to wait for Collectors
- Optimisation: Interactivity performance; pressing buttons doesn't lag anymore (#134)
- Bugfix: Accurate representation of currently running items
- Bugfix: Both family and families are represented in the GUI
- Bugfix: Plug-ins only compatible with the Context are more accurately represented. (#101)

Version 0.3.3
-------------

- Feature: Instance in exceptions (#106)
- Feature: Default on/off option for Plugins (#92)
- Feature: Hyperlink enabled documentation, errors and records. (#25)

Version 0.3.2
-------------

- Feature: Actions (#128)

Version 0.3.1
-------------

- Bugfix #122

Version 0.3.0
-------------

- Feature: Added button Validate (#115)
- Enhancement: Added the ability to customise the port
    at which QML distributes ports for connecting clients. (#114)
- Enhancement: Added additional metadata to terminal

Version 0.2.12
--------------
- Bugfix: GUI hangs on plug-ins subclassed from Plugin (#110)

Version 0.2.11
--------------

- Added support for Pyblish RPC 0.1.1
- PROBATION SUPPORT for pyblish_qml.settings
- PROBATION SUPPORT for context.set_data("label")

Version 0.2.10
--------------

- Feature: Talkback (see #95)
- Feature: Shared instance (see #96)
- Feature: Context visible in list of items
- Enhancement: Now supports Pyblish 1.1
- API: Added Pyblish.Graphs module
- Known issue: Inactive plug-ins appear to process

Version 0.2.9
-------------

- Feature: Perspective
- Enhancement: Uniform sizes in Terminal
- Bugfix: Support for filtering multiple logging levels in the Terminal
- Bugfix: Repair now repairs broken context (see #73)
- API: New model (see #81)
- API: Added Pyblish.Scrollbar
- API: Added Pyblish.Spacer

Version 0.2.8
-------------

- Feature: Repair
- Feature: Compatible plug-ins updates interactively as you toggle instances.
- Bugfix: Failing Selectors won't prohibit Extraction.
- Enhancement: You can't close while publishing.
- Enhancement: You can't crash the GUI. I dare you.
- API: Added util.register_vendor_libraries
- API: Added util.deregister_vendor_libraries
- API: Changed util.invoke -> util.async

Version 0.2.7
-------------

- Feature: Terminal 2.0
- Feature: Selectors are now part of the GUI
- Feature: Plug-in documentation is now (brutally) parsed
- Performance: GUI shows up before processing selectors.
- Bugfix: Plug-ins now only processes the Context once
- API: Added Pyblish.AwesomeIcon
- API: Added util.invoke

Version 0.2.6
-------------

- Process now properly dies when parent dies
- Known issue from  0.2.5 resolved; showing an already shown GUI
- State 1.0 (See #60)
- Refactoring

Version 0.2.5
-------------

- Preloading
- Performance enhancement (deferred requests)

Known issues:
    1. Showing an already visible GUI will cause the GUI
    to re-appear once closed.
    2. GUI may not appear gracefully under Windows using Aero
    or Windows 8, but may instead pop into place.

Version 0.2.4
-------------

- Implementing basic Properties page
- Implementing independent Endpoint server for testing
- Added Pyblish.TextArea
- Added Pyblish.ActionBar
- Added Pyblish.Action
- Added Pyblish.IconButton
- Improved feedback when clicking
- Improved feedback when hovering
- Made package executable
- Fixed issue 155
- Fixed issue 156
- Fixed issue 158
- Refactoring

Version 0.2.3
-------------

- Module consolidation (QML)

Version 0.2.2
-------------

- Refactoring
- Succeeded items now appear in green

Version 0.2.1
-------------

- Publishing now stops upon failed validation
- Errors are visualised
- Added Stop button
- Added Reset button

Version 0.2.0
-------------

- Swapped controller from QML to Python
- Performance enhancements; the GUI now rivals non-GUI modes.
- Simplified terminal output; it now resembles non-GUI output
- Simplified backend; no more thread-per-process
- Temporarily switched to native OS window; this should help GUI appearing behind windows
- Temporarily disabled pause/stop buttons; to stop, close the GUI.

Version 0.1.2
--------------

- Fixes black-font text sometimes appearing in GUI (#18)


pyblish_qml/host.py
import os
import sys
import time
import socket
import inspect
import traceback
import threading

# Python 2 and 3 compatibility
from .vendor.six.moves import xmlrpc_client as xmlrpclib

import pyblish.api

from . import rpc, settings


self = sys.modules[__name__]
self.ACTIVE_HOST_PORT = None
self.ACTIVE_PROXY = None
self.ACTIVE_HOST_NAME = None


def register_dispatch_wrapper(wrapper):
    """Register a dispatch wrapper for servers

    The wrapper must have this exact signature:
        (func, *args, **kwargs)

    """

    signature = inspect.getargspec(wrapper)
    if any([len(signature.args) != 1,
            signature.varargs is None,
            signature.keywords is None]):
        raise TypeError("Wrapper signature mismatch")

    rpc.server.dispatch_wrapper = wrapper


def deregister_dispatch_wrapper():
    rpc.server.dispatch_wrapper = None


def dispatch_wrapper():
    return rpc.server.dispatch_wrapper


def install(initial_port=9001):
    """Perform first time install

    Attributes:
        initial_port (int, optional): Port from which to start
            looking for available ports, defaults to 9001

    """

    if self.ACTIVE_HOST_PORT is not None:
        uninstall()

    install_host()
    install_callbacks()

    try:
        # In case QML is live and well, ask it
        # for the next available initial_port number.
        self.ACTIVE_PROXY = proxy()
        self.ACTIVE_HOST_PORT = self.ACTIVE_PROXY.find_available_port(
            initial_port)

    except (socket.timeout, socket.error):
        raise socket.error(
            "Is Pyblish QML running?\n\n"

            "This GUI requires a separate, standalone process to be running\n"
            "in the background in order to make use of it from within {host}\n"
            "Make sure to run python -m pyblish_qml prior to showing it from\n"
            "from {host}.\n\n"

            "See the documentation for further information.\n"
            "- https://github.com/pyblish/pyblish-qml\n\n#"
            .format(
                host=self.ACTIVE_HOST_NAME or "a host")
        )

    os.environ["PYBLISH_CLIENT_PORT"] = str(self.ACTIVE_HOST_PORT)

    try:
        _serve()

    except:
        exc_type, exc_value, exc_traceback = sys.exc_info()
        message = "".join(traceback.format_exception(
            exc_type, exc_value, exc_traceback))
        sys.stderr.write(message)
        sys.stderr.write("Setup failed..\n")


def uninstall():
    """Kill server and clean up traces of Pyblish QML in this process"""
    if self.ACTIVE_PROXY is None:
        return

    del self.ACTIVE_PROXY

    self.ACTIVE_HOST_PORT = None
    self.ACTIVE_PROXY = None

    uninstall_callbacks()
    uninstall_server()

    sys.stdout.write("Pyblish QML shutdown successful.\n")


def show():
    """Attempt to show GUI

    Requires install() to have been run first, and
    a live instance of Pyblish QML in the background.

    """

    if not self.ACTIVE_HOST_PORT:
        sys.stdout.write("Running install() for the first time.\n")
        install()

    if self.ACTIVE_PROXY is None:
        self.ACTIVE_PROXY = proxy()

    try:
        self.ACTIVE_PROXY.show(
            self.ACTIVE_HOST_PORT,
            settings.to_dict()
        )

    except (socket.error, socket.timeout):
        raise socket.error(
            "Communication problem.\n"
            "\n"
            "Hello? Is it me you're looking for?\n"
            "\n"
            "This problem typically occurs when Pyblish QML \n"
            "was running, but isn't anymore.\n"
            "\n"
            "To start it again, go to any terminal any type:\n"
            "$ python -m pyblish_qml"
        )


def install_callbacks():
    pyblish.api.register_callback("instanceToggled", _toggle_instance)


def uninstall_callbacks():
    pyblish.api.deregister_callback("instanceToggled", _toggle_instance)


def uninstall_server():
    rpc.server.kill()


def _serve():
    """Start XML-RPC server

    This is the entrypoint towards which the remote Pyblish QML instance
    communicates.

    """

    assert self.ACTIVE_HOST_PORT

    def server():
        """Provide QML with a friend to speak with"""
        self.server = rpc.server.start_production_server(self.ACTIVE_HOST_PORT)

    def heartbeat_emitter():
        """Let QML know we're still here"""
        p = proxy()

        while True:
            try:
                p.heartbeat(self.ACTIVE_HOST_PORT)
                time.sleep(1)
            except (socket.error, socket.timeout):
                pass

    for worker in (server, heartbeat_emitter):
        t = threading.Thread(target=worker, name=worker.__name__)
        t.daemon = True
        t.start()

    sys.stdout.write("Server running @ %i\n" % self.ACTIVE_HOST_PORT)

    return self.ACTIVE_HOST_PORT


def _toggle_instance(instance, new_value, old_value):
    """Alter instance upon visually toggling instance"""
    instance.data["publish"] = new_value


def proxy(timeout=5):
    """Return proxy at default location of Pyblish QML"""
    return xmlrpclib.ServerProxy(
        "http://127.0.0.1:9090",
        allow_none=True)


def install_host():
    """Install required components into supported hosts

    An unsupported host will still run, but may encounter issues,
    especially with threading.

    """

    for install in (_install_maya,
                    _install_houdini,
                    _install_nuke):
        try:
            install()
        except ImportError:
            pass
        else:
            break


def _install_maya():
    """Helper function to Autodesk Maya support"""
    from maya import utils

    def threaded_wrapper(func, *args, **kwargs):
        return utils.executeInMainThreadWithResult(
            func, *args, **kwargs)

    sys.stdout.write("Setting up Pyblish QML in Maya\n")
    register_dispatch_wrapper(threaded_wrapper)

    # Configure GUI
    settings.ContextLabel = "Maya"
    settings.WindowTitle = "Pyblish (Maya)"

    self.ACTIVE_HOST_NAME = "Maya"


def _install_houdini():
    """Helper function to SideFx Houdini support"""
    import hdefereval

    def threaded_wrapper(func, *args, **kwargs):
        return hdefereval.executeInMainThreadWithResult(
            func, *args, **kwargs)

    sys.stdout.write("Setting up Pyblish QML in Houdini\n")
    register_dispatch_wrapper(threaded_wrapper)

    settings.ContextLabel = "Houdini"
    settings.WindowTitle = "Pyblish (Houdini)"

    self.ACTIVE_HOST_NAME = "Houdini"


def _install_nuke():
    """Helper function to The Foundry Nuke support"""
    import nuke

    def threaded_wrapper(func, *args, **kwargs):
        return nuke.executeInMainThreadWithResult(
            func, args, kwargs)

    sys.stdout.write("Setting up Pyblish QML in Nuke\n")
    register_dispatch_wrapper(threaded_wrapper)

    settings.ContextLabel = "Nuke"
    settings.WindowTitle = "Pyblish (Nuke)"

    self.ACTIVE_HOST_NAME = "Nuke"
